// appdynamics/lib/probes/dynamodb-probe.js
/*
 * Copyright (c) AppDynamics, Inc., and its affiliates
 * 2016
 * All Rights Reserved
 * THIS IS UNPUBLISHED PROPRIETARY CODE OF APPDYNAMICS, INC.
 * The copyright notice above does not evidence any actual or intended publication of such source code
 */
'use strict';

function DynamoDbProbe(agent) {
  this.agent = agent;
  this.packages = ['aws-sdk'];
}
exports.DynamoDbProbe = DynamoDbProbe;

DynamoDbProbe.prototype.attach = function(obj) {
  var self = this;

  if(obj.__appdynamicsProbeAttached__) return;
  obj.__appdynamicsProbeAttached__ = true;
  self.agent.on('destroy', function() {
    if(obj.__appdynamicsProbeAttached__) {
      delete  obj.__appdynamicsProbeAttached__;
      proxy.release(obj.DynamoDB);
    }
  });

  var proxy = self.agent.proxy;
  var profiler = self.agent.profiler;

  proxy.after(obj, 'DynamoDB', function(obj, args, ret) {
    var dynamodb = ret,
      supportedProperties = {
        'VENDOR': 'Amazon',
        'SERVICE': 'DynamoDB',
        'ENDPOINT': dynamodb.endpoint.href
      };

    proxy.before(dynamodb, 'makeRequest', function(obj, args) {
      var command = args.length > 0 ? args[0] : undefined,
        params = (args.length > 1 && typeof(args) === 'object') ? args[1] : undefined,
        time = profiler.time();

      var exitCall = profiler.createExitCall(time, {
        exitType: 'EXIT_CUSTOM',
        exitSubType: 'Amazon Web Services',
        backendName: 'Dynamo DB',
        configType: 'Aws',
        supportedProperties: supportedProperties,
        command: profiler.sanitize(command),
        commandArgs: profiler.sanitize(JSON.stringify(params)),
        stackTrace: profiler.stackTrace()
      });

      if(!exitCall) return;

      var hasCallback = proxy.callback(args, -1, function(obj, args) {
        if(!time.done()) return;

        if(exitCall) {
          var error = proxy.getErrorObject(args);
          profiler.addExitCall(time, exitCall, error);
        }
      });

      if(exitCall && !hasCallback) {
        if(time.done()) {
          profiler.addExitCall(time, exitCall);
        }
      }
    }, undefined, true);
  }, true);
};
