// appdynamics/lib/probes/ioredis-probe.js
/*
Copyright (c) AppDynamics, Inc., and its affiliates
2015
All Rights Reserved
 */
'use strict';


function IoredisProbe(agent) {
  this.agent = agent;

  this.packages = ['ioredis'];
}
exports.IoredisProbe = IoredisProbe;



IoredisProbe.prototype.attach = function(obj) {
  var self = this;

  if(obj.__appdynamicsProbeAttached__) return;
  obj.__appdynamicsProbeAttached__ = true;
  self.agent.on('destroy', function() {
    if(obj.__appdynamicsProbeAttached__) {
      delete obj.__appdynamicsProbeAttached__;
      proxy.release(obj.Cluster.prototype.sendCommand);
      proxy.release(obj.prototype.sendCommand);
    }
  });

  var proxy = self.agent.proxy;
  var profiler = self.agent.profiler;
  var clusters = {};


  proxy.before(obj.Cluster.prototype, "sendCommand", function(obj) {
    var serverPool = [];
    if(Array.isArray(obj.startupNodes)) {
      obj.startupNodes.forEach(function(node) {
        var address = node.host + ':' + node.port;
        serverPool.push(address);
        clusters[address] = serverPool;
      });
    }
  });


  proxy.before(obj.prototype, "sendCommand", function(obj, args) {
    var redis = obj;
    var command = args[0];
    var time = profiler.time();
    var commandName = command.name;
    var commandArgs = command.args;
    var address = redis.options.host + ':' + redis.options.port;

    var serverPool = clusters[address];
    if(serverPool) {
      address = serverPool.join('\n');
    }

    var supportedProperties = {
      'SERVER POOL': address,
      'VENDOR': 'REDIS'
    };

    var exitCall = profiler.createExitCall(time, {
      exitType: 'EXIT_CACHE',
      backendName: 'Redis',
      supportedProperties: supportedProperties,
      command: commandName,
      commandArgs: profiler.sanitize(commandArgs),
      stackTrace: profiler.stackTrace()
    });

    if (!exitCall) return;

    if(command.callback && typeof(command.callback) === 'function') {
      proxy.before(command, 'callback', function(obj, args) {
        if(!time.done()) return;

        var error = proxy.getErrorObject(args);

        if(exitCall) {
          profiler.addExitCall(time, exitCall, error);
        }
      });

      proxy.before(command.promise, '_promise0', function(obj, args) {
        if(!time.done()) return;

        var error = proxy.getErrorObject(args);

        if(exitCall) {
          profiler.addExitCall(time, exitCall, error);
        }
      });
    }
    else if(exitCall) {
      if(time.done()) {
        profiler.addExitCall(time, exitCall);
      }
    }
  });
};
