// appdynamics/lib/probes/memcached-probe.js
/*
Copyright (c) AppDynamics, Inc., and its affiliates
2015
All Rights Reserved
 */
'use strict';


var commands = [
  'get',
  'gets',
  'getMulti',
  'set',
  'replace',
  'add',
  'cas',
  'append',
  'prepend',
  'increment',
  'decrement',
  'incr',
  'decr',
  'del',
  'delete',
  'version',
  'flush',
  'samples',
  'slabs',
  'items',
  'flushAll',
  'statsSettings',
  'statsSlabs',
  'statsItems',
  'cachedump'
];


function MemcachedProbe(agent) {
  this.agent = agent;

  this.packages = ['memcached'];
}
exports.MemcachedProbe = MemcachedProbe;



MemcachedProbe.prototype.attach = function(obj) {
  var self = this;

  if(obj.__appdynamicsProbeAttached__) return;
  obj.__appdynamicsProbeAttached__ = true;
  self.agent.on('destroy', function() {
    if(obj.__appdynamicsProbeAttached__) {
      delete obj.__appdynamicsProbeAttached__;
      commands.forEach(function(command) {
        proxy.release(obj[command]);
      });
    }
  });

  var proxy = self.agent.proxy;
  var profiler = self.agent.profiler;

  commands.forEach(function(command) {
    proxy.before(obj.prototype, command, function(obj, args) {
      // ignore, getMulti will be called
      if(command === 'get' && Array.isArray(args[0])) return;
      var client = obj;
      var params = args;

      var time = profiler.time();
      var serverPool = self.getServerPool(client.servers);

      var supportedProperties = {
        'SERVER POOL': serverPool.join('\n'),
        'VENDOR': 'MEMCACHED'
      };

      var exitCall = profiler.createExitCall(time, {
        exitType: 'EXIT_CACHE',
        backendName: 'Memcached',
        supportedProperties: supportedProperties,
        command: command,
        commandArgs: profiler.sanitize(params),
        stackTrace: profiler.stackTrace()
      });

      if (!exitCall) return;

      var hasCallback = proxy.callback(args, -1, function(obj, args) {
        if (exitCall) {
          if(!time.done()) return;

          if(client.servers.length === 0) {
                              // incomplete exit call
            return;
          }

          var error = proxy.getErrorObject(args);

          profiler.addExitCall(time, exitCall, error);
        }});

      if(exitCall && !hasCallback) {
        if(time.done()) {
          profiler.addExitCall(time, exitCall);
        }
      }
    });
  });
};

MemcachedProbe.prototype.getServerPool = function(servers) {
  var serverPool = [];

  if(typeof(servers) === 'string') {
    serverPool = [servers];
  }
  else if(typeof(servers) === 'object') {
    if(Array.isArray(servers)) {
      servers.forEach(function(server) {
        serverPool.push(server);
      });
    }
    else {
      for(var prop in servers) {
        serverPool.push(prop);
      }
    }
  }
  serverPool.sort();

  return serverPool;
};
